ChildProject/pipelines/sampler.py

```diff
@@ -79,6 +79,79 @@ def add_parser(samplers):
         parser.add_argument('--target-speaker-type', help = 'speaker type to get chunks from', choices=['CHI', 'OCH', 'FEM', 'MAL'], nargs = '+', default = ['CHI'])
         parser.add_argument('--sample-size', help = 'how many samples per recording', required = True, type = int)
 
+import numpy as np
+from scipy.fft import fft
+from pydub import AudioSegment
+
+class EnergyDetection(Sampler):
+    def __init__(self,
+        project: ChildProject.projects.ChildProject,
+        windows_length: float,
+        windows_spacing: float,
+        windows_count: int,
+        threshold: float = None,
+        lowpass_freq: int = -1,
+        initial_segments: str = None
+        ):
+
+        super().__init__(project)
+        self.windows_length = windows_length
+        self.windows_count = windows_count
+        self.windows_spacing = windows_spacing
+        self.threshold = threshold
+        self.lowpass_freq = lowpass_freq
+        self.initial_segments_path = initial_segments
+
+    def compute_energy_loudness(self, chunk, sampling_frequency: int):
+        chunk_fft = fft([x/(2**15 - 1) for x in chunk.get_array_of_samples()])
+
+        if self.lowpass_freq > 0:
+            chunk_fft = chunk_fft[:int(len(chunk)*self.lowpass_freq/sampling_frequency)]
+        
+        return np.sum(np.abs(chunk_fft)**2)/len(chunk)
+
+    def get_recording_windows(self, profile, recording):
+        audio = AudioSegment.from_wav(os.path.join(self.project.path, ChildProject.projects.ChildProject.RAW_RECORDINGS, recording['filename']))
+        duration = audio.duration_seconds
+        frequency = int(audio.frame_rate)
+
+        windows_starts = (1000*np.arange(self.windows_spacing, duration - self.windows_length, self.windows_spacing)).astype(int)
+        windows = []
+
+        for start in windows_starts:
+            energy = self.compute_energy_loudness(audio[start:start+int(1000*self.windows_length)], frequency)
+
+            windows.append({
+                'segment_onset': start/1000,
+                'segment_offset': start/1000+self.windows_length,
+                'recording_filename': recording['filename'],
+                'energy': energy
+            })
+
+        return windows
+        
+
+    def sample(self):
+        segments = []
+        for recording in self.project.recordings.to_dict(orient = 'records'):
+            windows = pd.DataFrame(self.get_recording_windows('', recording))
+            threshold = windows['energy'].quantile(0.1)
+            windows = windows[windows['energy'] >= threshold]
```

Comment:
so, currently, we exclude the 10% of windows we the lowest energy, and sample from the rest. We could use much higher quantiles, e.g. 80% if we really want loudness. This avoids the use of arbitrary thresholds which is problematic for a number of reasons:
 - amplitudes are defined up to some factor, i.e. audio levels are relative, so how to define a universel threshold ?
 - it is difficult to make sense of some energy threshold, BUT it is much easier to make sense of a percentile based threshold. If we assume loud = speech, then, the threshold could be compared to the expected fraction of speech time of the recordings...

Reply:
Agree with what has been said above ! 

If I follow you, the pipeline looks like this :
1) Exclude 80 % of windows w/ the lowest energies (which window length ? which time step (i.e. how much overlap between 2 successive windows ?))
2) Sample 10, 20 or N fixed-length windows amongst the 20% remaining windows.

Step 2) is important as we may want to be able to extract the same number of windows for 2 datasets whose cumulated duration is different (just using the quantile threshold doesn't allow us to do that).

For 2), I'd just use a uniform sampling : what do you think ? I think we can't skip the "try and listen" approach here. 
Don't hesitate to send me some extracted hunks :)

Reply:
> Agree with what has been said above !
> 
> If I follow you, the pipeline looks like this :
> 
> 1. Exclude 80 % of windows w/ the lowest energies (which window length ? which time step (i.e. how much overlap between 2 successive windows ? are those hard-coded ?))
> 2. Sample 10, 20 or N fixed-length windows amongst the 20% remaining windows.
> 
> Step 2) is important as we may want to be able to extract the same number of windows for 2 datasets whose cumulated duration is different (just using the quantile threshold doesn't allow us to do that).

You are right. Regarding the parameters, they are options:

```bash
$ child-project sampler . energy-detection --help
usage: child-project sampler path energy-detection [-h] --windows-length
                                                   WINDOWS_LENGTH
                                                   --windows-spacing
                                                   WINDOWS_SPACING
                                                   --windows-count
                                                   WINDOWS_COUNT
                                                   [--threshold THRESHOLD]
                                                   [--low-freq LOW_FREQ]
                                                   [--high-freq HIGH_FREQ]

optional arguments:
  -h, --help            show this help message and exit
  --windows-length WINDOWS_LENGTH
                        length of each window (in seconds)
  --windows-spacing WINDOWS_SPACING
                        spacing between windows (in seconds)
  --windows-count WINDOWS_COUNT
                        how many windows so sample from
  --threshold THRESHOLD
                        lowest energy quantile to sample from. default is 0.8
                        (i.e., sample from the 20% windows with the highest
                        energy).
  --low-freq LOW_FREQ   remove all frequencies below low-freq before
                        calculating each window's energy. (in Hz)
  --high-freq HIGH_FREQ
                        remove all frequencies above high-freq before
                        calculating each window's energy. (in Hz)
```

e.g.:

```bash
child-project sampler examples/valid_raw_data/ energy-detection \ 
    --windows-length 30 \
    --windows-spacing 120 \
    --windows-count 10 \
     --low-freq 100 \
     --high-freq 4000 \
     --threshold 0.9 \
     out.csv
```
 
> For 2), I'd just use a uniform sampling : what do you think ? I think we can't skip the "try and listen" approach here.
> Don't hesitate to send me some extracted hunks :)
> 
> EDIT : 2) is for the general case (when we don't want to compare between 2 recorders)

2 is indeed a uniform sampling. 
I will extract chunks. Here is what I can do : extract 5 chunks randomly for each energy decile. Is that good for you ?

Reply:
This will be very slow if you use it on a whole dataset...

```python
import os
import pandas as pd
from pydub import AudioSegment

from ChildProject.projects import ChildProject
from ChildProject.pipelines.sampler import EnergyDetectionSampler

project = ChildProject("examples/valid_raw_data")
project.read()

sampler = EnergyDetectionSampler(project, 30, 120, 500, 0)
sampler.sample()
sampler.segments.sort_values('energy', inplace = True)
sampler.segments['pos'] = 1
sampler.segments['decile'] = (10*(sampler.segments['pos'].cumsum()-1)/sampler.segments.shape[0]).astype(int)

chunks = sampler.segments.groupby('decile').sample(5)

for chunk in chunks.to_dict(orient = 'records'):
    recording_path = os.path.join(project.path, ChildProject.RAW_RECORDINGS, chunk['recording_filename'])
    audio = AudioSegment.from_wav(recording_path)

    onset = int(chunk['segment_onset']*1000)
    offset = int(chunk['segment_offset']*1000)

    print(os.path.join('windows', str(chunk['decile']), '{}_{}.mp3'.format(onset, offset)))

    os.makedirs(os.path.join('windows', str(chunk['decile'])), exist_ok = True)
    audio[onset:offset].export(os.path.join('windows', str(chunk['decile']), '{}_{}.mp3'.format(onset, offset)), format = 'mp3')
```

Reply:
not sure if this will be helpful, but there's a slither of data human-coded from children who wore the previous babylogger+LENA (at the same time). It sounds like these data could provide a human gold standard. I don't know where the annotations or audios are -- Xuan Nga oversaw them, and they were used for the pygamma examples in the seshat paper.

Reply:
Manel promised me some BabyLogger + LENA data, I guess that was it ? I'll ask her!